ocaml/xapi/tar_ext.ml

```diff
@@ -0,0 +1,118 @@
+let dir_perm = 0o755
+
+let dir_size = 4096L
+
+let ( + ) = Int64.add
+
+exception Illegal_file_path of string
+
+exception Unsupported_file_type of string
+
+exception File_exceeds_max_size_limit of string
+
+exception File_incomplete
+
+exception File_parsed_failed
+
+let unpack dir acc_size max_size ifd head =
+  let assert_file_path_is_legal file_path =
+    if
+      List.exists
+        (fun s -> String.equal "." s || String.equal ".." s)
+        (String.split_on_char '/' file_path)
+    then
+      raise (Illegal_file_path file_path)
+  in
+  let assert_file_not_exceed_max_size acc_size =
+    debug "Current total unpacked file size: %Ld" acc_size ;
+    if Int64.compare acc_size max_size > 0 then
+      raise
+        (File_exceeds_max_size_limit
+           ("File exceeds max size limit: " ^ Int64.to_string max_size)
+        )
+  in
+  let file_name = head.Tar.Header.file_name in
+  debug "Unpacking tar file: %s" file_name ;
+  assert_file_path_is_legal file_name ;
+  let file_path = dir ^ file_name in
+  match head.Tar.Header.link_indicator with
+  | Tar.Header.Link.Normal ->
+      let file_size = head.Tar.Header.file_size in
+      let new_acc_size = acc_size + file_size in
+      Unixext.with_file file_path [Unix.O_WRONLY; Unix.O_CREAT; Unix.O_TRUNC]
+        file_perm (fun ofd ->
+          assert_file_not_exceed_max_size new_acc_size ;
+          Unixext.copy_file ?limit:(Some file_size) ifd ofd |> ignore
+      ) ;
+      new_acc_size
+  | Tar.Header.Link.Directory ->
+      let new_acc_size = acc_size + dir_size in
+      assert_file_not_exceed_max_size new_acc_size ;
+      Unixext.mkdir_rec file_path dir_perm ;
+      new_acc_size
+  | Hard
+  | Symbolic
+  | Character
+  | Block
+  | FIFO
+  | GlobalExtendedHeader
+  | PerFileExtendedHeader
+  | LongLink ->
+      raise
+        (Unsupported_file_type
+           (Tar.Header.Link.to_string head.Tar.Header.link_indicator)
+        )
+
+(* Unpack a tar file to specific directory including following parameters:
+   dir: unpacking destination directory.
+   ifd: the descriptor of the file to be unpacked.
+   max_size: the total size limitation of unpacked files. Unit: Byte.
+*)
+let unpack_tar_file dir ifd max_size =
+  debug "Unpacked to %s" dir ;
+  Unixext.rm_rec dir ;
+  Unixext.mkdir_rec dir dir_perm ;
+  let rec unpack_all_files unpack dir acc_size max_size =
+    let acc_size =
+      Tar_unix.Archive.with_next_file ifd (unpack dir acc_size max_size)
+    in
+    unpack_all_files unpack dir acc_size max_size
+  in
+  try unpack_all_files unpack dir 0L max_size with
+  | Tar.Header.End_of_stream ->
+      debug "Unpacked successfully" ;
+      ()
+  | e -> (
+      error "Unpacked failed with exception: %s" (Printexc.to_string e) ;
+      match e with
+      | Illegal_file_path _
+      | Unsupported_file_type _
+      | File_exceeds_max_size_limit _ ->
+          raise e
+      | End_of_file | Tar.Header.Checksum_mismatch ->
+          raise File_incomplete
+      | _ ->
```

Comment:
catch all exceptions at this level are best avoided, or if we use them then we have to log them, and/or include them in the exception that we reraise.
Otherwise all the Unix errors here will be lost, which will make failures harder to debug.